### PR TITLE
BUG: Fix lobpcg() largest option

### PR DESCRIPTION
**Pull request by rc**
This PR applies fixes proposed by @lobpcg to lobpcg(), and addresses #4174.

### CODE REVIEW

File: scipy/sparse/linalg/eigen/lobpcg/lobpcg.py
Comment (tylerjereddy) [inline scipy/sparse/linalg/eigen/lobpcg/lobpcg.py]: Just to be sure--the fact that the math here is rather challenging justifies retaining an unused function as a comment? I see the mathematician providing this solution in the linked issue has suggested to comment out.

Comment (rc) [reply scipy/sparse/linalg/eigen/lobpcg/lobpcg.py]: This is a work-in-progress solution -  the lines will be either removed or a warning will be issued. @lobpcg wants to collect more data on why np.float32 matrices sometimes fail, see the linked issue comment.

Comment (ilayn) [reply scipy/sparse/linalg/eigen/lobpcg/lobpcg.py]: You can add "WIP:" to the title to make it a bit more obvious

Comment (rc) [reply scipy/sparse/linalg/eigen/lobpcg/lobpcg.py]: Done, thanks for the tip!

**Comment by lobpcg** [comment]
These bug fixes are needed to pass the tests in https://github.com/scikit-learn/scikit-learn/pull/12319 
@ilayn Is there a tentative plan/timeline for the actual merge of this PR, so that I plan my scikit-learn contribution accordingly?

**Comment by ilayn** [comment]
It is currently WIP so only would attract interested people. Is this ready for review?

Regardless, I think scikit-learn is following the official releases and not the master branch. Hence it will have to wait the release in any case.

**Comment by lobpcg** [comment]
It is WIP only because there are a few lines commented out, for some potential future attention. If it is a problem for the review, these lines can of course be just removed, or the tolerance values chosen so large that the 2 assertions always trivially pass. Yet another alternative is to put these 2 assertions  inside 
   if verbosityLevel > 10: 

scikit-learn test
https://ci.appveyor.com/project/sklearn-ci/scikit-learn/builds/19323001/job/f30rwvudp8vu2wyn
ran today uses 
https://files.pythonhosted.org/packages/c4/f3/752fd6778a9d07fddb2b02dac5895287e594d2d0d156a2a422c710f6a851/scipy-1.1.0-cp37-none-win_amd64.whl

Not sure what it actually means.

**Comment by rc** [comment]
Yes, I think it is ready, except resolving what to do with those now commented-out symmetry checks.

**Comment by rc** [comment]
I have removed the WIP status.

**Comment by ilayn** [comment]
I don't know if it is relevant in the sparse case but in the dense case, the symmetry check is often done via 

```python
import numpy as np
from scipy.linalg import norm

# Mat is the supposedly hermitian matrix
Antisym = Mat - Mat.conj().T

if norm(Antisym, 1) > max(np.spacing(10**a), (10**b)*norm(Mat, 1)):
    raise ValueError('not symmetric enough')
```
where `a, b` are typically around 3 and -1. These quantities are, in my personal experience, tricky to tune via `rtol` and `atol` since they are dependent on the matrix data. 

If it is of critical importance it has to be added in and tuned to avoid future maintenance burden, otherwise better be taken out to reduce the clutter.

**Comment by rc** [comment]
Thanks, @ilayn.

So I have replaced the `_assert_symmetric()` with new `_report_nonhermitian()`, that gets called for a nonzero verbosity, and only reports the nonhermitian gram matrices, instead of raising an exception. In this way, a user might get more info in case of non-convergence. I have also created a new test `test_verbosity()` that call `lobpcg()` with non-zero verbosity levels. For the moment, it just uses the elastic rod matrices.

What do you think?

**Comment by lobpcg** [comment]
@rc To be able to control and fix all the errors, I have made a local copy at https://github.com/lobpcg/scikit-learn/blob/lobpcg_svd/sklearn/utils/lobpcg.py for PR https://github.com/scikit-learn/scikit-learn/pull/12319 

We should try to keep in full sync with the original
     from scipy.sparse.linalg import lobpcg

LOBPCG in sklearn is used to solve normal eigenvalue problems X'*X as called from https://github.com/lobpcg/scikit-learn/blob/lobpcg_svd/sklearn/utils/extmath.py function lobpcg_svd
It goes through multiple tests performing PCA and truncated_svd, and has multiple failures, I am fixing. It appears that  Travis CI and continuous-integration/appveyor checks may follow stricter standards there (e.g., lines 79 vs, 80 characters long), giving many formatting errors from lobpcg.py, which I am fixing. In these fixes, I have deviated quite a bit now from the original, which is bad. And I am still getting formatting errors from lobpcg.py so the editing is not yet finished... Could you please diff this original vs. https://github.com/lobpcg/scikit-learn/blob/lobpcg_svd/sklearn/utils/lobpcg.py and try to merge so we have a single version that passes all tests both here and at sklearn? My original did not include the latest changes you have made in this PR, so they need to be merged. I solved for the smallest of -X'*X in lobpcg_svd to go around the trouble with the largest. 
 
If you agree with this plan, please put WIP label back here, since merging may take some time and having a local copy of lobpcg at sklearn removes the urgency of this PR. 

What do you think?

**Comment by rc** [comment]
Yes, this code in SciPy was written well before the automatic style checking with CI tools and I was, IIRC, following the style/naming conventions you used in the matlab version - many things need to be style-corrected to adhere to PEP8.

However, I would not mix the style update with this PR - it is IMHO better to have several smaller PRs than one large do-it-all kind - it is certainly easier for reviewers. So, if you agree, I would port this PR fixes to your version in scikit-learn, and I would not mark this WIP to get it merged as soon as we get the green light? The style and other updates then may come in other PRs.

We also need to discuss with the SciPy developers and maintainers your broader plans on making the generic LOBPCG algorithm version with various backends - let us create a new issue for that?

**Comment by lobpcg** [comment]
If small increments are better, we should surely put this PR through as is now, pending only requested review changes. 

For the next PR, I would propose making a PEP8 etc compliant lobpcg.py and may be adding more tests (I have not yet looked at scipy lobpcg tests...) I hope that I am getting close to PEP8 etc compliant lobpcg.py in  https://github.com/lobpcg/scikit-learn/blob/lobpcg_svd/sklearn/utils/lobpcg.py for PR scikit-learn/scikit-learn#12319 If you agree, it would be grate if you could help with it, since some of the errors are way above my beginner Python knowledge. Having this done, we would have a solid base to discuss with the SciPy developers and maintainers the broader plans on making the generic LOBPCG algorithm version with various backends.

**Comment by rc** [comment]
OK. I have just forked scikit-learn - I will try fixing remaining style errors tomorrow. Let's move further discussion to https://github.com/scikit-learn/scikit-learn/pull/12319.

**Comment by lobpcg** [comment]
> OK. I have just forked scikit-learn - I will try fixing remaining style errors tomorrow. Let's move further discussion to [scikit-learn/scikit-learn#12319](https://github.com/scikit-learn/scikit-learn/pull/12319).

@rc Great, thanks! I stop making changes in lobpcg there, waiting for your fixes. Please merge the changes made for this PR there - I have not.

**Comment by rc** [comment]
@ilayn, @tylerjereddy, could you review, please? We have made additional improvements to the LOBPCG solver, but it is better to create a new focused PR instead of adding them to this one, right?

File: scipy/sparse/linalg/eigen/lobpcg/lobpcg.py
Comment (ilayn) [inline scipy/sparse/linalg/eigen/lobpcg/lobpcg.py]: I made a mistake here this line should probably be

```python
tol = np.spacing(max(10**a, (10**b)*norm(M, 1)))
```

I am typing on the phone so I can't see all at once but the idea is simply to take the maximum of `1000` and `some scaling * norm of the matrix` and take the np.spacing of it. 

However I don't know what does print statements do. Are you supposed to give a warning there?

File: scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py
Comment (ilayn) [inline scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py]: This can be done via 
```python
vals = np.arange(1, n+1, dtype=float)
```
Similarly for n=5 below

File: scipy/sparse/linalg/eigen/lobpcg/lobpcg.py
Comment (rc) [reply scipy/sparse/linalg/eigen/lobpcg/lobpcg.py]: I will fix `tol` (if this function survives, see below), thanks!

It is not really a warning, this function is only meant to print info on potential non-symmetry when the verbosity level is non-zero. I have added it instead of the commented-out assertions. I am not sure about its usefulness, though, @lobpcg?

File: scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py
Comment (rc) [reply scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py]: Thanks, that is what careless copy-pasting without reading does :)

File: scipy/sparse/linalg/eigen/lobpcg/lobpcg.py
Comment (lobpcg) [reply scipy/sparse/linalg/eigen/lobpcg/lobpcg.py]: In principle, it is useful to cheaply check the symmetry, especially since the matrices of the eigenproblem to solve can be passed as user-defined functions... 

It is probably best to use one of the already existing functions, such as
https://docs.scipy.org/doc/numpy/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose 

The specific values of rtol and atol are tricky to choose well, since they depend on the data type, float32 vs foat64, the matrix size, and the algorithm used for the comparison... However, it is not so practically important now, since we currently put the check only for a non-default verbosityLevel. 

@rc whatever changes you finally make here, please remember to sync with https://github.com/lobpcg/scikit-learn/blob/lobpcg_svd/sklearn/utils/lobpcg.py and check that it does not lead to formatting errors.

Comment (rc) [reply scipy/sparse/linalg/eigen/lobpcg/lobpcg.py]: Yes, choosing rtol and atol is tricky, that is why @ilayn proposed the way above.

**Comment by rc** [comment]
I have tried to address the review comments, let us know what more needs to be done to get this PR merged.

FYI: @lobpcg and me have more updates and significant fixes coming (now living in https://github.com/scikit-learn/scikit-learn/pull/12319), that build on the fixes here.

File: scipy/sparse/linalg/eigen/lobpcg/lobpcg.py
Comment (tylerjereddy) [inline scipy/sparse/linalg/eigen/lobpcg/lobpcg.py]:
```suggestion
        print('matrix %s is not sufficiently Hermitian for a=%d, b=%d:'
```

File: scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py
Comment (tylerjereddy) [inline scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py]: This could be a separate unit test for clarity maybe

File: scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py
Comment (tylerjereddy) [inline scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py]: so we're not asserting anything here--nothing gets printed because the matrices are sufficiently Hermitian?

**Review by tylerjereddy — COMMENTED** [review]
I added a few more minor comments -- maybe @ilayn can take one more look.

@pv opened the original issue -- if I'm not mistaken this solution is slightly different from the early-stage proposal of deprecating `largest`. Is that ok with you Pauli?

I see that scikit-learn is depending on the resolution of this PR. Good things here are the apparently minimal disruptions to any old tests, the fact that the CI is all green, and apparently a few experienced mathematicians have commented either here / in the linked issue.

**Comment by pv** [comment]
If I understand correctly, this PR changes what `lobpcg` returns if the `largest` option is not specified. If that's right, I'd suggest changing the default value to `largest=False` (which was effectively the previous default value) --- the results probably will be in different order, but at least from the same end of the spectrum as previously.

Can you also add `largest=False` to `test_eigs_consistency`. Granted, it's tested in the other tests, but those consider more complex test matrices.

**Comment by lobpcg** [comment]
@pv PR fixes both issues in #4174 for "largest=True": 
 1. finding the correct, largest, eigenvalues for all matrix sizes - a mistake in the code, now fixed. 
 2. outputting the largest in a "natural", descending order, just as in the ARPACK solver

The current unchanged default "largest=True" is also the default in ARPACK, making it consistent with lobpcg. 

There is no change for `largest=False` that returns the smallest in ascending order, also consistent with ARPACK.  Full consistency of lobpcg and ARPACK minimizes user confusion and makes simple writing comparative tests, e.g., already coded for PCA and partial SVD in scikit-learn/scikit-learn#12319

> If I understand correctly, this PR changes what `lobpcg` returns if the `largest` option is not specified. If that's right, I'd suggest changing the default value to `largest=False` (which was effectively the previous default value) --- the results probably will be in different order, but at least from the same end of the spectrum as previously.
> 
> Can you also add `largest=False` to `test_eigs_consistency`. Granted, it's tested in the other tests, but those consider more complex test matrices.

**Comment by pv** [comment]
Do as you see best. However, the backward incompatible change in what
`lobpcg(A, x)` returns (larges eigenvalues by default, not the
smallest) must be mentioned in release notes, i.e. added to this page
after this PR is merged:
https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.2.0

File: scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py
Comment (rc) [reply scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py]: You mean the alternative branch test? The alternative branch means here the small matrix code path in lobpcg().

File: scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py
Comment (rc) [reply scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py]: The idea behind this test was to run the code paths that are not normally run in the default (silent) operation - there are other print statements in lobpcg(). So yes, we are not asserting anything. And actually, for the current settings of a=3, b=-1 in _report_nonhermitian() calls the gramA matrix is reported non-Hermitian. This could be fixed either by tweaking the parameters, or removing the _report_nonhermitian() completely. A cleaner solution how to provide optional diagnostics to users might be to support callbacks at suitable places in lobpcg() - what do you think?

File: scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py
Comment (tylerjereddy) [reply scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py]: Great, maybe add that as a comment if you split off to a second test or use parametrization--the two tests repeat a lot of the same logic.

File: scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py
Comment (tylerjereddy) [reply scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py]: I'm not sure I now enough about the issue to ask for any substantive additional changes--I might politely suggest maybe just asserting that you get an expected result when probing the verbosity handling, but maybe not crucial.

File: scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py
Comment (tylerjereddy) [inline scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py]: The lines are identical apart from the `verbosityLevel`. I'm tempted to suggest parametrization here too, but I know some people don't like that unless there are > 2 cases.

File: scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py
Comment (rc) [reply scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py]: Done!

File: scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py
Comment (rc) [reply scipy/sparse/linalg/eigen/lobpcg/tests/test_lobpcg.py]: I have changed the verbosity level to 11 to make sure really all related code is executed. Subsequently, I have fixed two bugs. The code runs now without errors and this is the expected result - I am not sure how to assert that in the test? :)

**Comment by rc** [comment]
@tylerjereddy I have now some spare time to work on this - let me know if there are some additional issues.

**Review by tylerjereddy — APPROVED** [review]
Ok, this has been through a few rounds of review and there's a fair bit of expert commenting.

CI is all green & Pauli's backward-compatibility concerns seem to be addressed.

Thanks @rc and @lobpcg

**Comment by rc** [comment]
@tylerjereddy Thanks for merging, I have added a paragraph to the release notes.

**Comment by rc** [comment]
#4174 can be closed now.

**Comment by ilayn** [comment]
done

**Pull request merged**
